maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenTestsPlugin.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
+
+import aQute.bnd.osgi.Builder;
+import aQute.bnd.osgi.Constants;
+
+@Mojo(name = "bnd-process-tests", defaultPhase = LifecyclePhase.PROCESS_TEST_CLASSES, requiresDependencyResolution = ResolutionScope.COMPILE, threadSafe = true)
+public class BndMavenTestsPlugin extends AbstractBndMavenPlugin {
+
+	/**
+	 * If true, make the tests artifact of fragment of the main artifact.
+	 */
+	@Parameter(defaultValue = "false")
+	private boolean									fragment;
+
+	/**
+	 * Possible values are {@link TestCases#junit3 junit3},
+	 * {@link TestCases#junit4 junit4}, {@link TestCases#junit5 junit5},
+	 * {@link TestCases#all all} and {@link TestCases#useTestCasesHeader}
+	 */
+	@Parameter(defaultValue = "junit5")
+	private TestCases								testCases;
+
+	@Parameter(defaultValue = "${project.build.testSourceDirectory}", readonly = true)
+	private File									sourceDir;
+
+	@Parameter(defaultValue = "${project.build.testResources}", readonly = true)
+	private List<org.apache.maven.model.Resource>	resources;
+
+	@Parameter(defaultValue = "${project.build.outputDirectory}", readonly = true)
+	private File									mainClassesDir;
+
+	@Parameter(defaultValue = "${project.build.testOutputDirectory}", readonly = true)
+	private File									classesDir;
+
+	@Parameter(defaultValue = "${project.build.testOutputDirectory}", readonly = true)
+	private File									outputDir;
+
+	@Parameter(defaultValue = "${project.build.testOutputDirectory}/META-INF/MANIFEST.MF", readonly = true)
+	private File									manifestPath;
+
+	@Parameter(property = "bnd-tests.skip", defaultValue = "false")
+	private boolean									skip;
+
+	@Override
+	public File getSourceDir() {
+		return sourceDir;
+	}
+
+	@Override
+	public List<Resource> getResources() {
+		return resources;
+	}
+
+	@Override
+	public File getClassesDir() {
+		return classesDir;
+	}
+
+	@Override
+	public File getOutputDir() {
+		return outputDir;
+	}
+
+	@Override
+	public File getManifestPath() {
+		return manifestPath;
+	}
+
+	@Override
+	public boolean isSkip() {
+		return skip;
+	}
+
+	@Override
+	protected void processBuilder(Builder builder) throws MojoFailureException {
+		String bsn = builder.getProperty(Constants.BUNDLE_SYMBOLICNAME);
+		if (bsn == null) {
+			bsn = project.getArtifactId();
+			builder.setProperty(Constants.BUNDLE_SYMBOLICNAME, bsn + "-tests");
+		}
+
+		if (testCases != TestCases.useTestCasesHeader) {
+			builder.setProperty(Constants.TESTCASES, "${sort;${uniq;" + testCases.filter() + "}}");
+		} else if (builder.getProperty(Constants.TESTCASES) == null) {
+			throw new MojoFailureException("<testCases> specified testCasesHeader but no Test-Cases header was found");
+		}
+
+		if (fragment) {
+			builder.setProperty(Constants.FRAGMENT_HOST, bsn);
```

Comment:
This is wrong here if the test goal configuration sets a BSN and has fragment=true since you make the bundle a fragment of itself. You need to know the BSN of the other goal's bundle. Perhaps the fragment configuration parameter should not be a boolean but a string which is the bsn of the host? Setting fragment to `${project.artifactId}` is probably the choice when you want your test bundle to be a fragment.

Reply:
fixed.